helper/constraints/semver/constraints_test.go

```diff
@@ -0,0 +1,125 @@
+package semver
+
+import (
+	"testing"
+
+	"github.com/hashicorp/go-version"
+)
+
+// This file is a copy of github.com/hashicorp/go-version/constraint_test.go
+// with minimal changes to demonstrate differences. Diffing the files should
+// illustrate behavior differences in Constraint and version.Constraint.
+
+func TestNewConstraint(t *testing.T) {
+	cases := []struct {
+		input string
+		count int
+		err   bool
+	}{
+		{">= 1.2", 1, false},
+		{"1.0", 1, false},
+		{">= 1.x", 0, true},
+		{">= 1.2, < 1.0", 2, false},
+
+		// Out of bounds
+		{"11387778780781445675529500000000000000000", 0, true},
+
+		// Semver only
+		{">= 1.0beta1", 0, true},
+
+		// No pessimistic operator
+		{"~> 1.0", 0, true},
+	}
+
+	for _, tc := range cases {
+		v, err := NewConstraint(tc.input)
+		if tc.err && err == nil {
+			t.Fatalf("expected error for input: %s", tc.input)
+		} else if !tc.err && err != nil {
+			t.Fatalf("error for input %s: %s", tc.input, err)
+		}
+
+		if len(v) != tc.count {
+			t.Fatalf("input: %s\nexpected len: %d\nactual: %d",
+				tc.input, tc.count, len(v))
+		}
+	}
+}
+
+func TestConstraintCheck(t *testing.T) {
+	cases := []struct {
+		constraint string
+		version    string
+		check      bool
+	}{
+		{">= 1.0, < 1.2", "1.1.5", true},
+		{"< 1.0, < 1.2", "1.1.5", false},
+		{"= 1.0", "1.1.5", false},
+		{"= 1.0", "1.0.0", true},
+		{"1.0", "1.0.0", true},
+
```

Comment:
We have an open issue https://github.com/hashicorp/nomad/issues/4729 where the documented behavior:

> The ordering is compared lexically.

Doesn't match what you'd naively expect for numeric comparisons (and we want both a bug fix and doc fix for). Maybe it'd be a good idea to have a test here that exercises lexical sorting, which will demonstrate the documented behavior is a little more sophisticated than "lexical" at least for some operators:

```go
{"> 10", "8", false},
```

Reply:
Good call! Added and pushed.